### ISSUE TITLE
Encoder fused SDPA 3.6× slower at L=512 after #4259 (e02b78ba → 5e715ea)

### ISSUE THREAD

**Issue by GOavi101** [torch-spyre/torch-spyre#4339]
## Summary

`torch.compile` of fused (unmasked) `F.scaled_dot_product_attention` got ~3.6× slower between:

- **fast:** `e02b78ba35f9a1a69a458c3149e9c01d9f4fa6a8`
- **slow:** `5e715ea70c4aa9e76e45b7fbae6b92a6d7658764` (#4259 SDPA rewrite, 34 commits later)

Isolated to the **fused** kernel. Packed QK and P·V on the same shape/dtype are unchanged ( on both SHAs). Repro is a single `torch.compile` of permute + non-causal `F.sdpa` — no spyre-inference.

## When it shows up

Compiling encoder self-attention (granite-embedding-125m) on Spyre:

- `B=1`, `H=12`, `L=512`, `D=64`, fp16
- non-causal `F.sdpa`, **no `attn_mask`**
- layout is `[L, H, D]` → permute to `[1, H, L, D]` then fused SDPA (the B=1 fused path used when the prompt fills the L=512 bucket)

The slowdown appears as soon as that fused kernel is compiled and run on `5e715ea`. The same compile on `e02b78ba` is fast. Packed (masked) QK / P·V at the same shape are not affected.

## Isolated microbench

Shape `[512, 12, 64]` fp16, Spyre, `torch.compile(dynamic=False)`:

#4259 dropped the old `work_div` tiling. At L=512 the new decomp uses one 512-wide Q tile. Forcing tile=64 on the **new** decomp unrolls 8×8 online-softmax without the old `work_div` path and is worse, not a workaround.

## Repro

Run the same script against the two torch-spyre SHAs above. Wipe inductor cache when switching version or tile size.

```bash
rm -rf /tmp/torchinductor_* ~/.cache/torchinductor
python repro_fused_sdpa.py
```

Optional, **on `5e715ea` only** — shows shrinking the tile cap is not a fix:

```bash
rm -rf /tmp/torchinductor_* ~/.cache/torchinductor
python repro_fused_sdpa.py --sdpa-tile-size 64
```

## Script (`repro_fused_sdpa.py`)

```python
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import os
import statistics
import time

os.environ.setdefault("TORCH_DEVICE_BACKEND_AUTOLOAD", "0")
os.environ.setdefault("RANK", "0")
os.environ.setdefault("LOCAL_RANK", "0")
os.environ.setdefault("WORLD_SIZE", "1")
os.environ.setdefault("LOCAL_WORLD_SIZE", "1")

import torch
import torch.nn.functional as F


def fused_sdpa(q, k, v, scale):
    q = q.unsqueeze(0).transpose(1, 2)
    k = k.unsqueeze(0).transpose(1, 2)
    v = v.unsqueeze(0).transpose(1, 2)
    out = F.scaled_dot_product_attention(q, k, v, is_causal=False, scale=scale)
    t, h, d = q.shape[2], q.shape[1], q.shape[3]
    return out.transpose(1, 2).reshape(t, h, d)


def main() -> None:
    p = argparse.ArgumentParser()
    p.add_argument("--seq", type=int, default=512)
    p.add_argument("--heads", type=int, default=12)
    p.add_argument("--dim", type=int, default=64)
    p.add_argument("--warmup", type=int, default=3)
    p.add_argument("--iters", type=int, default=20)
    p.add_argument("--sdpa-tile-size", type=int, default=None)
    args = p.parse_args()

    import torch_spyre

    torch_spyre._autoload()

    tile = None
    try:
        import torch_spyre._inductor.decompositions as decomp

        tile = getattr(decomp, "_SDPA_MAX_SEQUENCE_TILE_SIZE", None)
        if args.sdpa_tile_size is not None and tile is not None:
            decomp._SDPA_MAX_SEQUENCE_TILE_SIZE = args.sdpa_tile_size
            tile = args.sdpa_tile_size
    except Exception:
        pass

    device = torch.device("spyre")
    dtype = torch.float16
    scale = args.dim ** -0.5
    gen = torch.Generator().manual_seed(0)
    q = torch.randn(args.seq, args.heads, args.dim, dtype=dtype, generator=gen).to(device)
    k = torch.randn(args.seq, args.heads, args.dim, dtype=dtype, generator=gen).to(device)
    v = torch.randn(args.seq, args.heads, args.dim, dtype=dtype, generator=gen).to(device)

    fn = torch.compile(fused_sdpa, dynamic=False)
    print(f"tile={tile}  [{args.seq},{args.heads},{args.dim}]")
    t0 = time.perf_counter()
    fn(q, k, v, scale)
    torch.spyre.synchronize()
    print(f"compile+first {time.perf_counter() - t0:.1f}s")

    for _ in range(args.warmup):
        fn(q, k, v, scale)
        torch.spyre.synchronize()
    times = []
    for _ in range(args.iters):
        torch.spyre.synchronize()
        t0 = time.perf_counter()
        fn(q, k, v, scale)
        torch.spyre.synchronize()
        times.append((time.perf_counter() - t0) * 1000)
    print(f"fused ms_mean={statistics.mean(times):.3f}  min={min(times):.3f}")


if __name__ == "__main__":
    main()
```

**Comment by ani300** [comment]
The regression is addressed by #4424.

#4259 replaced the former full-core work division with indiscriminate coarse tiling. The new selector uses work division only when the shape is placeable and its estimated live values fit within the frontend-managed LX budget; otherwise it retains the tiled fallback. For this repro (B=1, H=12, Lq=Lkv=512, D=64, fp16), it selects one full 512-wide block with work division H/Q/K = 4/8/8. Its estimated live footprint is 443,136 bytes/core, below the 1,625,344-byte frontend budget.

I reran the exact repro on the same machine with isolated compiler caches:

- Pre-regression e02b78ba: mean 0.415 ms, min 0.401 ms (3 warmups, 20 iterations)
- #4424 at 6613e870: mean 0.424 ms, min 0.416 ms (3 warmups, 20 iterations)

That is within about 2% on the mean. A longer 10-warmup/100-iteration run was noisy in its synchronized mean, but the fixed branch reached a lower minimum (0.341 ms versus 0.394 ms on e02b78ba). This indicates the 3.6x regression is removed and performance is back in the pre-regression range.

PR: https://github.com/torch-spyre/torch-spyre/pull/4424
